packages/core/src/rules/oas3/spec-components-invalid-map-name.ts

```diff
@@ -0,0 +1,53 @@
+import { Oas3Rule } from '../../visitors';
+import { Problem, UserContext } from '../../walk';
+import { Location } from '../../ref-utils';
+
+export const SpecComponentsInvalidMapName: Oas3Rule = () => {
+  const KEYS_REGEX = '^[a-zA-Z0-9.\\-_]+$';
+
+  function validateKey(
+    key: string | number,
+    report: (problem: Problem) => void,
+    location: Location,
+    component: string
+  ) {
+    if (!new RegExp(KEYS_REGEX).test(key as string)) {
+      report({
+        message: `The map key in ${component} "${key}" does not match the regular expression "${KEYS_REGEX}"`,
+        location: location.key(),
+      });
+    }
+  }
+
+  return {
+    Components: {
+      Parameter(_node, { key, report, location }: UserContext) {
+        validateKey(key, report, location, 'parameters');
+      },
+      Response(_node, { key, report, location }: UserContext) {
+        validateKey(key, report, location, 'responses');
+      },
+      Schema(_node, { key, report, location }: UserContext) {
+        validateKey(key, report, location, 'schemas');
+      },
+      Example(_node, { key, report, location }: UserContext) {
+        validateKey(key, report, location, 'examples');
+      },
+      RequestBody(_node, { key, report, location }: UserContext) {
+        validateKey(key, report, location, 'requestBodies');
+      },
+      Header(_node, { key, report, location }: UserContext) {
+        validateKey(key, report, location, 'headers');
+      },
+      SecurityScheme(_node, { key, report, location }: UserContext) {
+        validateKey(key, report, location, 'securitySchemas');
+      },
+      Link(_node, { key, report, location }: UserContext) {
+        validateKey(key, report, location, 'links');
+      },
+      Callback(_node, { key, report, location }: UserContext) {
+        validateKey(key, report, location, 'callbacks');
+      },
```

Comment:
It looks like it's better to use `any()` visitor and manually check whether we're inside of the `Components`.
Maybe we can go even more strict and require ALL keys to conform to this pattern? @adamaltman  do you think it makes sense?

Reply:
Well, look at this:

```yaml
    patternProperties:
      '^(schemas|responses|parameters|examples|requestBodies|headers|securitySchemes|links|callbacks|pathItems)$':
        $comment: Enumerating all of the property names in the regex above is necessary for unevaluatedProperties to work as expected
        propertyNames:
          pattern: '^[a-zA-Z0-9._-]+$'
```

I think we should check those (schemas, responses, parameters, examples, etc...). I'm not sure if that's `any()`, but it is a lot. Also, the examples map names don't only appear inside of `Components` but can be directly inside of operations too (probably the same is true with some other things).

Reply:
Well, yes, so it looks like this code misses this scenario where the map is not used in `Components`:

```yaml
openapi: 3.1.0
info: 
  title: Path examples bad map
  version: error-case
paths:
  /foo:
    get: 
      responses: 
        '200': 
          description: OK
          content: 
            application/json: 
              schema: 
                type: object
              examples:
                My invalid example:
                  value: 
                    foo: bar
                A_valid_example:
                  value: 
                    baz: quz
  
```

Reply:
Wow, this is not specified: 

https://github.com/OAI/OpenAPI-Specification/blob/main/schemas/v3.1/schema.yaml

Seems like a bug in the schema. I opened up this issue: https://github.com/OAI/OpenAPI-Specification/issues/3013

Reply:
@tatomyr It is related only to specific fields according specification https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#componentsObject

Reply:
Thanks for pointing to the spec @IgorKarpiuk. Then your code makes sense. 

The only open question that remains is whether this restriction should be imposed on `Components` only or not. 

Meanwhile, I think it's okay to go with the current approach.

Reply:
I agree. We can see or adjust it based on any responses here: https://github.com/OAI/OpenAPI-Specification/issues/3013